### PR TITLE
Refactoring analysis process in order to finish the next job like `constructor detection`

### PR DESCRIPTION
**Pull request by Basstorm**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

Currently, `check for vtables` is a separate task for analyzing classes. This is not good for analyzing other information about the classes, and in my humble opinion, it should be a sub-task of `check for classes`.
The output of the `acll` command should currently look like this, which includes other member functions of the class instead of just the virtual functions.
```
A
  (vtable at 0x400d28)
  method_A_A @ 0x400b36
  method_A_greet @ 0x400ac8 (vtable + 0x0)
  method_A_printValue @ 0x400af4 (vtable + 0x8)
B: A
  (vtable at 0x400d08)
  method_B_B_int @ 0x400b4e
  method_A_greet @ 0x400ac8 (vtable + 0x0)
  method_B_printValue @ 0x400b82 (vtable + 0x8)
C: A
  (vtable at 0x400ce8)
  method_C_C_int @ 0x400bc4
  method_A_greet @ 0x400ac8 (vtable + 0x0)
  method_C_printValue @ 0x400bf8 (vtable + 0x8)
std::basic_ostream_char__std::char_traits_char_____std::endl_char__std
  char_traits_char____std::basic_ostream_char__std::char_traits_char_____ @ 0x6012a8
std::basic_ostream_char__std::char_traits_char_____std::operator____std
  char_traits_char____std::basic_ostream_char__std::char_traits_char______char_const__ @ 0x601298
std::ios_base::Init
  Init__ @ 0x601278
  _Init__ @ 0x601290
std::ostream
  operator___int_ @ 0x601268
  operator___std::ostream______std::ostream___ @ 0x6012a0
```
One problem is that the function name of the class obtained from `rz-bin` contains `()~,`etc., such as `set_value(int, int)`, which causes the special symbols to be escaped to `_` after `rz_analysis_class_method_set`, which is hard to see in the output of `acll`, how can I just get its function name without other symbols?

**Test plan**

...

**Closing issues**

...

### CODE REVIEW

**Comment by wargio** [comment]
```
[XX] db/cmd/dwarf "big endian variable info"
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'e asm.bytes=true
e asm.dwarf=false
aaa
pdf @  dbg.init
' bins/elf/ppc64_sudoku_dwarf
-- stdout
@@ -1,6 +1,7 @@
             ; CALL XREF from dbg.main @ 0x10001ae8
             ;-- Grid::init():
             ;-- method.Grid.init:
+            ;-- method.Grid.method.Grid.init:
 / dbg.init (Grid * const this);
 |           ; arg Grid * const this @ r3
 |           0x100012bc      39030144       addi  r8, r3, 0x144         ; arg1 ; void init(Grid * const this);


[XX] db/cmd/cmd_fnj fnj shows demangled symbols
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'aaa
fj~{253}
fnj~{253}
' bins/elf/demangle-test-cpp
-- stdout
@@ -1,2 +1,2 @@
-{"name":"reloc.operator_delete_void","realname":"reloc.operator delete(void*)","size":8,"offset":16432}
-{"name":"reloc.operator_delete_void","realname":"reloc.operator delete(void*)","size":8,"offset":16432}
+{"name":"method.int__std.__copy_move_a_true__int___int___int___int___int","size":49,"offset":7747}
+{"name":"method.int__std.__copy_move_a_true__int___int___int___int___int","size":49,"offset":7747}
```

**Comment by wargio** [comment]
anyway, awesome job.

**Comment by Basstorm** [comment]
> ```
> [XX] db/cmd/dwarf "big endian variable info"
> RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'e asm.bytes=true
> e asm.dwarf=false
> aaa
> pdf @  dbg.init
> ' bins/elf/ppc64_sudoku_dwarf
> -- stdout
> @@ -1,6 +1,7 @@
>              ; CALL XREF from dbg.main @ 0x10001ae8
>              ;-- Grid::init():
>              ;-- method.Grid.init:
> +            ;-- method.Grid.method.Grid.init:
>  / dbg.init (Grid * const this);
>  |           ; arg Grid * const this @ r3
>  |           0x100012bc      39030144       addi  r8, r3, 0x144         ; arg1 ; void init(Grid * const this);
> 
> 
> [XX] db/cmd/cmd_fnj fnj shows demangled symbols
> RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'aaa
> fj~{253}
> fnj~{253}
> ' bins/elf/demangle-test-cpp
> -- stdout
> @@ -1,2 +1,2 @@
> -{"name":"reloc.operator_delete_void","realname":"reloc.operator delete(void*)","size":8,"offset":16432}
> -{"name":"reloc.operator_delete_void","realname":"reloc.operator delete(void*)","size":8,"offset":16432}
> +{"name":"method.int__std.__copy_move_a_true__int___int___int___int___int","size":49,"offset":7747}
> +{"name":"method.int__std.__copy_move_a_true__int___int___int___int___int","size":49,"offset":7747}
> ```

Thanks for reminding me, it's fixed now.

**Review by wargio — APPROVED** [review]
LGTM

**Review by Maijin — APPROVED** [review]

**Review by XVilka — APPROVED** [review]
Looks good, but lets wait for @thestr4ng3r look since he wrote much of the classes analysis.

**Comment by XVilka** [comment]
I think this PR is good as is, but we might want to reuse the hashtable for classes, since the amount of them can be huge, so RzList becomes inefficient. Thus, I think such optimization could be done in later PRs.

File: librz/analysis/class.c [hunk 1]
```diff
@@ -53,6 +53,25 @@ static const char *attr_type_id(RzAnalysisClassAttrType attr_type) {
...
 }
 
+RZ_API void rz_analysis_class_recover_from_rzbin(RzAnalysis *analysis) {
+	rz_cons_break_push(NULL, NULL);
+	RzList *classes = rz_bin_get_classes(analysis->binb.bin);
+	if (classes) {
+		RzListIter *iter_class;
+		RzBinClass *class;
+		rz_list_foreach (classes, iter_class, class) {
+			if (!rz_analysis_class_exists(analysis, class->name)) {
```
Comment (thestr4ng3r) [inline librz/analysis/class.c:66]: If you use `rz_cons_break_push()`, you should also check `rz_cons_is_breaked()` in every iterationa and break if it returns true.

File: librz/analysis/class.c [hunk 2]
```diff
@@ -478,13 +497,78 @@ static RzAnalysisClassErr rz_analysis_class_add_attr_unique(RzAnalysis *analysis
 // Format: addr,vtable_offset
 
 static char *flagname_method(const char *class_name, const char *meth_name) {
+	if (rz_str_startswith(meth_name, "method.")) {
+		return rz_str_new(meth_name);
```
Comment (thestr4ng3r) [inline librz/analysis/class.c:516]: In which practical case is this branch taken?

File: librz/analysis/class.c [hunk 3]
```diff
@@ -478,13 +497,78 @@ static RzAnalysisClassErr rz_analysis_class_add_attr_unique(RzAnalysis *analysis
...
+	}
+}
+
+RZ_API bool rz_analysis_class_method_exists(RzAnalysis *analysis, const char *class_name, const char *meth_name) {
+	char *content = rz_analysis_class_get_attr(analysis, class_name, RZ_ANALYSIS_CLASS_ATTR_TYPE_METHOD, meth_name, false);
+	if (!content) {
+		return false;
+	}
+
+	if (!rz_str_sanitize_sdb_key(meth_name)) {
```
Comment (thestr4ng3r) [inline librz/analysis/class.c]: Why this check here? I don't think it is related to checking if a method exists. Moreover, the returned pointer is leaked.

**Review by thestr4ng3r — CHANGES_REQUESTED** [review]
Overall quite nice, I like the results it produces by fusing the two data sources.

File: librz/analysis/class.c (hunk 1 above)
Comment (Basstorm) [reply librz/analysis/class.c:66]: okay, I see.

File: librz/analysis/class.c (hunk 3 above)
Comment (Basstorm) [reply librz/analysis/class.c]: Sorry, I copied it from `rz_analysis_class_method_get()` but I forgot to check what it does..

File: librz/analysis/class.c (hunk 2 above)
Comment (Basstorm) [reply librz/analysis/class.c:516]: In `rz_analysis_class_method_recover()`, the method name is from `rz_analysis_get_function_at()` when this function was already analyzed and added to the context, which contains `method.`. But `flagname_method()` which is called in `rz_analysis_class_method_set()` will add the prefix `method.` again.

**Comment by Basstorm** [comment]
I think there is still room for improvement, for example, the output method name should be its own method name instead of `method.classname.methodname`. Because the class inheritance relationship is already indicated in the output of the class name. And adding the name of the method itself to `RzAnalysisMethod` also facilitates further analysis work. What do you think? @thestr4ng3r @XVilka

**Comment by XVilka** [comment]
@Basstorm yes, this is a good point, it would be an improvement over the current output.

**Comment by XVilka** [comment]
Seems to break analysis serialization now:
```
55/76 serialize_analysis   FAIL            0.27s   killed by signal 11 SIGSEGV
56/76 serialize_config     OK              0.02s
57/76 serialize_flag       OK              0.02s
58/76 big                  OK              3.09s
59/76 serialize_spaces     OK              0.02s
60/76 sign                 OK              0.02s
61/76 skiplist             OK              0.02s
62/76 skyline              OK              0.02s
63/76 spaces               OK              0.02s
64/76 sparse               OK              0.02s
65/76 rbtree               OK              0.11s
66/76 stack                OK              0.02s
67/76 str                  OK              0.02s
68/76 strbuf               OK              0.02s
69/76 table                OK              0.02s
70/76 subprocess           OK              0.04s
71/76 task                 OK              0.02s
72/76 tree                 OK              0.02s
73/76 uleb128              OK              0.02s
74/76 unum                 OK              0.03s
75/76 util                 OK              0.02s
76/76 vector               OK              0.02s


The output from the failed tests:

55/76 serialize_analysis   FAIL            0.27s   killed by signal 11 SIGSEGV
--- command ---
06:16:55 /home/runner/work/rizin/rizin/build/test/unit/test_serialize_analysis
--- stderr ---
WARNING: rz_event_send: assertion 'ev && !ev->incall' failed (line 109)
WARNING: rz_event_send: assertion 'ev && !ev->incall' failed (line 109)
WARNING: rz_event_send: assertion 'ev && !ev->incall' failed (line 109)
WARNING: rz_event_send: assertion 'ev && !ev->incall' failed (line 109)
WARNING: rz_event_send: assertion 'ev && !ev->incall' failed (line 109)
WARNING: rz_event_send: assertion 'ev && !ev->incall' failed (line 109)
-------


Summary of Failures:

55/76 serialize_analysis   FAIL            0.27s   killed by signal 11 SIGSEGV


Ok:                 75  
Expected Fail:      0   
Fail:               1   
Unexpected Pass:    0   
Skipped:            0   
Timeout:            0   
```

**Comment by Basstorm** [comment]
It looks better now..
```
Polygon
  (vtable at 0x3d20)
  Polygon @ 0x11ea
  set_values @ 0x1496
  ~Polygon @ 0x121e
  area @ 0x14c0 (vtable + 0x0)
  sides @ 0x14d4 (vtable + 0x8)
Rectangle: Polygon
  (vtable at 0x3d00)
  ~Rectangle @ 0x1586
  Rectangle @ 0x1558
  area @ 0x14e8 (vtable + 0x0)
  sides @ 0x1508 (vtable + 0x8)
Triangle: Polygon
  (vtable at 0x3ce0)
  Triangle @ 0x15b4
  ~Triangle @ 0x15e2
  area @ 0x151c (vtable + 0x0)
  sides @ 0x1544 (vtable + 0x8)
std::basic_ostream_char__std::char_traits_char_____std::endl_char__std
  char_traits<char> > @ 0x3fd0
std::ios_base::Init
  Init @ 0x3fb0
  ~Init @ 0x3ff8
std::ostream
  operator<< @ 0x3fb8
```

File: librz/analysis/class.c (hunk 2 above)
Comment (thestr4ng3r) [reply librz/analysis/class.c:516]: Ok, looks good to me then.

**Review by thestr4ng3r — APPROVED** [review]

**Pull request merged**